### PR TITLE
`ultralytics 8.0.114` automatic optimizer selection

### PR DESCRIPTION
**Pull request by glenn-jocher**
### Summary
📝🚀🎉

This pull request updates the documentation, the pre-commit hook, the package version, and the optimizer logic for the YOLOv8 model. The main purpose of these changes is to enhance the user experience, the model performance, and the code quality of the ultralytics package.

> _We are the masters of the dark_
> _We train our models with the best `optimizer`_
> _We document our tasks with user-images_
> _We log our results with Comet and TensorBoard_

### Walkthrough
*  Update the version number and the pyupgrade hook to reflect the latest changes and improvements in the codebase ([link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceL3-R3), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9L25-R25))
*  Modify the default value and the logic of the optimizer argument to automatically select the best optimizer based on the number of iterations and the number of classes, and introduce a new equation for the learning rate for the AdamW optimizer to optimize the model performance ([link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-a21b6f90c80bba2698f9a8e30b1a002078a425376e5eb2864f078e0a97d32a3bL97-R97), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-9d692d4ad15d420a92105fcd432fb1d2221b71d5af055bea3bd2b70523ce6743L23-R23), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-1dc4dec7a1716e080109cc732dc44c6e843b30bcdd324e0141e679ebf718c2b1L630-R633), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-4a4a3691b08dc4cb045c88d4405e25fe3e154c35ddf71370e94e336b5f3b417dL75-R75))
*  Add a new section to the documentation on how to use different loggers to track, compare, and visualize the model's performance over time, and provide instructions on how to install, initialize, and use them ([link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-4a4a3691b08dc4cb045c88d4405e25fe3e154c35ddf71370e94e336b5f3b417dL105-R160))
*  Replace the image sources in the documentation from raw GitHub links to user-images links, or to different user-images links that show more relevant and clear examples, to improve the loading and display of the images ([link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-a1a4bd3ddfbcef91068c9a46f93c16283474fb8bae9aed652e5999df7fb12abcL9-R9), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-597c0ff446990af3bc46d0ec8bf12cd047c250472b8155ba60aa2dbc6eaec6f3L8-R8), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-315c549028b862e841c5597869bc1b00ccc48aa4494bc2adb459bc87ad6a3d7aL11-R11), [link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-6c5aa10a7e1f72b933f0873902a94468bd6cc633d9f5ada357bf50f7266807c3L9-R9))
*  Add a line to the greetings.yml workflow file to invite new contributors to join the Ultralytics Discord community for real-time conversations and collaborations ([link](https://github.com/ultralytics/ultralytics/pull/3037/files?diff=unified&w=0#diff-6318f5fc1ff2829ebbfd7846021a98d807a3b866fcfac669187f49fa22a40702R35-R36))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to optimize training and streamline contribution workflow.

### 📊 Key Changes
- Revised the default optimizer to `'auto'` for better automatic selection.
- Updated documentation images for various tasks.
- Bumped version to `8.0.114`.
- Improved the Python code upgrade hook by upgrading `pyupgrade` tool to `v3.4.0`.
- Added detailed documentation on logging with Comet, ClearML, and TensorBoard.
- Fine-tuned the automatic optimizer selection criteria based on the number of classes and iterations.

### 🎯 Purpose & Impact
These changes:
- 🛠️ **Enhance Training**: Auto-selection of optimizer potentially increases model performance without manual tweaking.
- 📚 **Improve Documentation**: Updated images make documentation clearer; logging instructions provide better experiment tracking.
- 🔧 **Update Tools**: Keeping tools like `pyupgrade` up-to-date ensures code quality and consistency.
- 💻 **Improve User Experience**: Simplifying the configuration process helps both new and existing users get started more easily.

### CODE REVIEW

**Comment by amm1111** [comment]
@glenn-jocher   hi, what is the principle behind this formula and where does it come from? about the auto lr_fit :                                                                    if name == 'auto':
            nc = getattr(model, 'nc', 10)  # number of classes
            lr_fit = round(0.002 * 5 / (4 + nc), 6)  # lr0 fit equation to 6 decimal places
            name, lr, momentum = ('SGD', 0.01, 0.9) if iterations > 10000 else ('AdamW', lr_fit, 0.9)

**Comment by glenn-jocher** [comment]
It’s not from a paper—`lr_fit = 0.002 * 5 / (4 + nc)` is an empirical heuristic we fit from internal sweeps to pick a *safe* starting `lr0` when `optimizer=auto`, scaling the LR down as `nc` (number of classes) grows to reduce early-training instability on multi-class heads; you can see exactly where it’s applied (and that `optimizer=auto` will override user `lr0`/`momentum`) in [`BaseTrainer.build_optimizer()`](https://docs.ultralytics.com/reference/engine/trainer/). If it doesn’t match your dataset/training length, just set `optimizer=` and `lr0=` explicitly, or run built-in tuning as described in the [Hyperparameter Tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/).
